### ISSUE TITLE
//tensorflow/python/kernel_tests:conv_ops_test fails on AVX512 builds

### ISSUE THREAD

**Issue by markdryan** [tensorflow/tensorflow#21265]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: No
- **TensorFlow installed from (source or binary)**: Source
- **TensorFlow version (use command below)**: commit ecd8decac3d9f3c7cd772e1561b9c2d3f23aa830 on the master branch
- **Python version**: Python 2.7.12
- **Bazel version (if compiling from source)**: [bazel release 0.15.0]
- **GCC/Compiler version (if compiling from source)**: 5.4.0
- **CUDA/cuDNN version**: n/a
- **GPU model and memory**: n/a
- **Exact command to reproduce**:

bazel test --config=opt --cache_test_results=no  -- //tensorflow/python/kernel_tests:conv_ops_test

### Describe the problem

The //tensorflow/python/kernel_tests:conv_ops_test test case consistently fails on AVX512 builds of tensorflow

### Source code / logs
```
======================================================================
FAIL: testConv2D3x3FilterStride1x1Same (__main__.DeepConv2DTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/user/.cache/bazel/_bazel_markus/59a925eeb655b30b5c683f8317fe569e/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/conv_ops_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/conv_ops_test.py", line 1796, in testConv2D3x3FilterStride1x1Same
    self._RunTestCases([1, 1], "SAME")
  File "/home/user/.cache/bazel/_bazel_markus/59a925eeb655b30b5c683f8317fe569e/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/conv_ops_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/conv_ops_test.py", line 1790, in _RunTestCases
    self._CompareFwdConv2D(input_shape, filter_shape, conv_strides, padding)
  File "/home/user/.cache/bazel/_bazel_markus/59a925eeb655b30b5c683f8317fe569e/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/conv_ops_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/conv_ops_test.py", line 1782, in _CompareFwdConv2D
    self.assertAllClose(values_expect, values_test, rtol=1e-5, atol=1e-5)
  File "/home/user/.cache/bazel/_bazel_markus/59a925eeb655b30b5c683f8317fe569e/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/conv_ops_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 1381, in assertAllClose
    self._assertAllCloseRecursive(a, b, rtol=rtol, atol=atol, msg=msg)
  File "/home/user/.cache/bazel/_bazel_markus/59a925eeb655b30b5c683f8317fe569e/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/conv_ops_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 1331, in _assertAllCloseRecursive
    (path_str, path_str, msg))
  File "/home/user/.cache/bazel/_bazel_markus/59a925eeb655b30b5c683f8317fe569e/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/conv_ops_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 1286, in _assertArrayLikeAllClose
    a, b, rtol=rtol, atol=atol, err_msg=msg, equal_nan=True)
  File "/home/user/.local/lib/python2.7/site-packages/numpy/testing/nose_tools/utils.py", line 1396, in assert_allclose
    verbose=verbose, header=header, equal_nan=equal_nan)
  File "/home/user/.local/lib/python2.7/site-packages/numpy/testing/nose_tools/utils.py", line 779, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-05, atol=1e-05
Mismatched value: a is different from b. 
(mismatch 96.60625%)
 x: array([[[[[1257.2019, 1264.4261, 1258.311 , ..., 1258.0616, 1228.0815,
           1245.9379],
          [1875.4814, 1885.9357, 1868.8992, ..., 1865.7366, 1842.2179,...
 y: array([[[[[1256.2858, 1263.7555, 1257.8694, ..., 1256.4175, 1226.5343,
           1244.2854],
          [1874.3364, 1884.7977, 1868.6038, ..., 1864.8274, 1841.5417,...

----------------------------------------------------------------------
Ran 73 tests in 5.034s

FAILED (failures=1)
not close where =  (array([0, 0, 0, ..., 0, 0, 0]), array([0, 0, 0, ..., 4, 4, 4]), array([0, 0, 0, ..., 4, 4, 4]), array([0, 0, 0, ..., 2, 2, 2]), array([  0,   1,   2, ..., 125, 126, 127]))
not close lhs =  [1257.2019 1264.4261 1258.311  ... 1840.9    1845.5197 1843.9106]
not close rhs =  [1256.2858 1263.7555 1257.8694 ... 1839.0669 1843.0771 1841.3849]
not close dif =  [0.9161377 0.6706543 0.4416504 ... 1.8331299 2.442505  2.5257568]
not close tol =  [0.01257286 0.01264755 0.01258869 ... 0.01840067 0.01844077 0.01842385]
dtype = float32, shape = (1, 5, 5, 5, 128)
```

```
FAIL: testConv2D3x3FilterStride1x1Valid (__main__.DeepConv2DTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/user/.cache/bazel/_bazel_markus/59a925eeb655b30b5c683f8317fe569e/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/conv_ops_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/conv_ops_test.py", line 1793, in testConv2D3x3FilterStride1x1Valid
    self._RunTestCases([1, 1], "VALID")
  File "/home/user/.cache/bazel/_bazel_markus/59a925eeb655b30b5c683f8317fe569e/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/conv_ops_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/conv_ops_test.py", line 1790, in _RunTestCases
    self._CompareFwdConv2D(input_shape, filter_shape, conv_strides, padding)
  File "/home/user/.cache/bazel/_bazel_markus/59a925eeb655b30b5c683f8317fe569e/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/conv_ops_test.runfiles/org_tensorflow/tensorflow/python/kernel_tests/conv_ops_test.py", line 1782, in _CompareFwdConv2D
    self.assertAllClose(values_expect, values_test, rtol=1e-5, atol=1e-5)
  File "/home/user/.cache/bazel/_bazel_markus/59a925eeb655b30b5c683f8317fe569e/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/conv_ops_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 1381, in assertAllClose
    self._assertAllCloseRecursive(a, b, rtol=rtol, atol=atol, msg=msg)
  File "/home/user/.cache/bazel/_bazel_markus/59a925eeb655b30b5c683f8317fe569e/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/conv_ops_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 1331, in _assertAllCloseRecursive
    (path_str, path_str, msg))
  File "/home/user/.cache/bazel/_bazel_markus/59a925eeb655b30b5c683f8317fe569e/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/conv_ops_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 1286, in _assertArrayLikeAllClose
    a, b, rtol=rtol, atol=atol, err_msg=msg, equal_nan=True)
  File "/home/user/.local/lib/python2.7/site-packages/numpy/testing/nose_tools/utils.py", line 1396, in assert_allclose
    verbose=verbose, header=header, equal_nan=equal_nan)
  File "/home/user/.local/lib/python2.7/site-packages/numpy/testing/nose_tools/utils.py", line 779, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-05, atol=1e-05
Mismatched value: a is different from b. 
(mismatch 94.3576388889%)
 x: array([[[[[2802.1404, 2796.723 , 2792.0303, ..., 2813.227 , 2780.2148,
           2796.4958],
          [2797.6106, 2792.4019, 2813.2893, ..., 2790.9302, 2788.8625,...
 y: array([[[[[2802.357 , 2796.4634, 2792.4202, ..., 2812.9595, 2780.5022,
           2796.4705],
          [2797.8032, 2792.7297, 2813.6294, ..., 2791.9949, 2789.6606,...

----------------------------------------------------------------------
Ran 73 tests in 7.059s

FAILED (failures=1)
not close where =  (array([0, 0, 0, ..., 0, 0, 0]), array([0, 0, 0, ..., 4, 4, 4]), array([0, 0, 0, ..., 2, 2, 2]), array([0, 0, 0, ..., 1, 1, 1]), array([  0,   1,   2, ..., 125, 126, 127]))
not close lhs =  [2802.1404 2796.723  2792.0303 ... 2763.8906 2751.459  2762.4377]
not close rhs =  [2802.357  2796.4634 2792.4202 ... 2764.6665 2753.163  2764.2102]
not close dif =  [0.21655273 0.25952148 0.38989258 ... 0.7758789  1.7041016  1.7724609 ]
not close tol =  [0.02803357 0.02797463 0.0279342  ... 0.02765667 0.02754163 0.0276521 ]
dtype = float32, shape = (1, 5, 3, 3, 128)
```

**Comment by markdryan** [comment]
The basic problem seems to be that the gemm_pack_rhs<...>::operator()
(https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/eigen_spatial_convolutions.h#L657)
doesn't always produce the correct output if the depth is not a multiple of the
packet_size.  Looking at the code there seem to be a number of issues, each of which are caught by
assertions.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/eigen_spatial_convolutions.h#L718
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/eigen_spatial_convolutions.h#L371

I've created a [patch](https://bitbucket.org/eigen/eigen/pull-requests/443/fix-tensor-contraction-for-avx512-machines/diff) for Eigen that fixes the issue by modifying TensorContractionBlocking to ensure the depth passed to gemm_pack_rhs<...>::operator() is always a multiple of the packet_size for packet_sizes > 8.  I appreciate Eigen might not be the best place to fix this issue as the issue is not reproducible with the Eigen Tensor class.  However, the multiple assertions in with gemm_pack_rhs<...>::operator() in Tensorflow led me to believe that this code was not intended to work when depth is not a multiple of the packet_size and so I opted to make the change in Eigen.  The change is localised to tensor contraction and is the simplest fix that I could come up with. If a fix in Tensorflow or elsewhere would be preferable, please let me know and I'll rework my patch.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 75 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 90 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 105 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 120 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by markdryan** [comment]
The Eigen patch has been updated to extend the fix to cover the new parallel contraction code.

https://bitbucket.org/eigen/eigen/pull-requests/554/fix-tensor-contraction-on-avx512-builds/diff

**Comment by markdryan** [comment]
The patch has now been merged into Eigen and commit #be4b5c3 pulls this patch into TensorFlow.  CNNs now work on AVX512 builds.  This issue can be closed.
